contracts/lens/QuoterV3.sol

```diff
@@ -0,0 +1,361 @@
+    address public v2Factory;
+
+    /// @dev Transient storage variable used to check a safety condition in exact output swaps.
+    uint256 private amountOutCached;
+
+    constructor(
+        address _factory,
+        address _v2Factory,
+        address _WETH9
+    ) PeripheryImmutableState(_factory, _WETH9) {
+        v2Factory = _v2Factory;
+    }
+
+    function getPool(
+        address tokenA,
+        address tokenB,
+        uint24 fee
+    ) private view returns (IUniswapV3Pool) {
+        return IUniswapV3Pool(PoolAddress.computeAddress(factory, PoolAddress.getPoolKey(tokenA, tokenB, fee)));
+    }
+
+    /**
+        @dev used for exactIn
+        @notice Given an amountIn, fetch the reserves of the V2 pair and call getAmountOut
+        @notice new addition
+     */
+    function getPairAmountOut(
+        uint256 amountIn,
+        address tokenIn,
+        address tokenOut
+    ) private view returns (uint256) {
+        (uint256 reserveIn, uint256 reserveOut) = UniswapV2Library.getReserves(v2Factory, tokenIn, tokenOut);
+        return UniswapV2Library.getAmountOut(amountIn, reserveIn, reserveOut);
+    }
+
+    /**
+        @dev used for exactOut
+        @notice Given an amountOut, fetch the reserves of the V2 pair and call getAmountIn
+        @notice new addition
+     */
+    function getPairAmountIn(
+        uint256 amountOut,
+        address tokenIn,
+        address tokenOut
+    ) private view returns (uint256) {
+        (uint256 reserveIn, uint256 reserveOut) = UniswapV2Library.getReserves(v2Factory, tokenIn, tokenOut);
+        return UniswapV2Library.getAmountIn(amountOut, reserveIn, reserveOut);
+    }
+
+    /// @inheritdoc IUniswapV3SwapCallback
+    function uniswapV3SwapCallback(
+        int256 amount0Delta,
+        int256 amount1Delta,
+        bytes memory path
+    ) external view override {
+        require(amount0Delta > 0 || amount1Delta > 0); // swaps entirely within 0-liquidity regions are not supported
+        (address tokenIn, address tokenOut, uint24 fee) = path.decodeFirstPool();
+        CallbackValidation.verifyCallback(factory, tokenIn, tokenOut, fee);
+
+        (bool isExactInput, uint256 amountToPay, uint256 amountReceived) =
+            amount0Delta > 0
+                ? (tokenIn < tokenOut, uint256(amount0Delta), uint256(-amount1Delta))
+                : (tokenOut < tokenIn, uint256(amount1Delta), uint256(-amount0Delta));
+
+        IUniswapV3Pool pool = getPool(tokenIn, tokenOut, fee);
+        (uint160 sqrtPriceX96After, int24 tickAfter, , , , , ) = pool.slot0();
+
+        if (isExactInput) {
+            assembly {
+                let ptr := mload(0x40)
+                mstore(ptr, amountReceived)
+                mstore(add(ptr, 0x20), sqrtPriceX96After)
+                mstore(add(ptr, 0x40), tickAfter)
+                revert(ptr, 96)
+            }
+        } else {
+            // if the cache has been populated, ensure that the full output amount has been received
```

Comment:
since no output swaps should be coming here we should revert with an error message